vm/systemSmartContracts/staking.go

```diff
@@ -88,16 +113,206 @@ func (r *stakingSC) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCod
 		return r.jail(args)
 	case "unJail":
 		return r.unJail(args)
+	case "changeRewardAddress":
+		return r.changeRewardAddress(args)
+	case "changeValidatorKeys":
+		return r.changeValidatorKey(args)
 	}
 
 	return vmcommon.UserError
 }
 
+func getPercentageOfValue(value *big.Int, percentage float64) *big.Int {
+	x := new(big.Float).SetInt(value)
+	y := big.NewFloat(percentage)
+
+	z := new(big.Float).Mul(x, y)
+
+	op := big.NewInt(0)
+	result, _ := z.Int(op)
+
+	return result
+}
+
+func (r *stakingSC) calculateStakeAfterBleed(startRound uint64, endRound uint64, stake *big.Int) *big.Int {
+	if startRound > endRound {
+		return stake
+	}
+	if endRound-startRound < r.numRoundsWithoutBleed {
+		return stake
+	}
+
+	totalRoundsToBleed := endRound - startRound - r.numRoundsWithoutBleed
+	totalPercentageToBleed := float64(totalRoundsToBleed) * r.bleedPercentagePerRound
+	totalPercentageToBleed = math.Max(totalPercentageToBleed, r.maximumPercentageToBleed)
+
+	bleedValue := big.NewInt(0).Neg(getPercentageOfValue(stake, totalPercentageToBleed))
```

Comment:
This is a bit awkward. What should I understand from a negative bleedValue? Is a reward? :D
I would have the bleed value positive here and subtract it from stake on the next line

Reply:
forgot about Sub.